tests/pri_queue_tests.cpp

```diff
@@ -0,0 +1,165 @@
+#include <thread>
+#include <boost/test/unit_test.hpp>
+#include <appbase/application.hpp>
+
+using namespace appbase;
+
+BOOST_AUTO_TEST_SUITE(pri_queues)
+
+constexpr static int q0 = appbase::default_queue;
+
+// start app thread and call execution loop based on use_default_queue flag 
+std::thread start_app_thread(appbase::scoped_app& app, bool use_default_queue) {
+   const char* argv[] = { boost::unit_test::framework::current_test_case().p_name->c_str() };
+   BOOST_CHECK(app->initialize(sizeof(argv) / sizeof(char*), const_cast<char**>(argv)));
+   app->startup();
+   std::thread app_thread( [&, use_default_queue]() {
+      if ( use_default_queue ) {
+         app->exec();
+      } else {
+         app->multi_queue_exec();
+      }
+   } );
+   return app_thread;
+}
+
+// wait for time out or the number of results has reached
+void wait_for_results(const std::vector<int>& results, size_t expected) {
+   auto i = 0;
+   while ( i < 10 && results.size() < expected) {
+      std::this_thread::sleep_for(std::chrono::milliseconds(1));
+      ++i;
+   }
+}
+
+// test single default queue
+BOOST_AUTO_TEST_CASE( default_single_queue ) {
+   // start app thread with default queue
+   appbase::scoped_app app;
+   auto app_thread = start_app_thread(app, true);
+
+   // post functions. each function pushs a value into the results queue
+   std::vector<int> results;
+   app->post( priority::medium, [&results]() { results.emplace_back(1); } );
+   app->post( priority::medium, [&results]() { results.emplace_back(2); } );
+   app->post( priority::high,   [&results]() { results.emplace_back(3); } );
+   app->post( priority::lowest, [&results]() { results.emplace_back(4); } );
+   app->post( priority::low,    [&results]() { results.emplace_back(5); } );
+   app->post( priority::low,    [&results]() { results.emplace_back(6); } );
+
+   // wait until app thread has executed all the functions
+   auto num_rslts_expected = 6;
+   wait_for_results(results, num_rslts_expected);
+   app->quit();
+   app_thread.join();
+   
+   // make sure functions are executed in right order
```

Comment:
At this point it would be safe to access `std::vector`, so you could keep these checks for order but determine test done via an `std::atomic<int>` total.

However, there is no guarantee that the first `medium` post is not executed before the `high` since it is posted first. It could run before the `high` is posted. It would be valid for `results[0] == 1`. For this particular test the only thing you guaranteed is that `3` executes before `4,5,6` and that `1` executes before `2` and `5` executes before `6`.

Reply:
Thanks for this! Will update all the tests.